### DISCUSSION TITLE
post/asio-recording-in-naudio

### DISCUSSION THREAD

**Discussion by giscus** [markheath/soundcode-blog-comments#40] [Announcements]
# post/asio-recording-in-naudio

Mark Heath's Development Blog

https://markheath.net/post/asio-recording-in-naudio

**Comment by dansted** [comment]
Hi, I've tried to apply an effect to the input signal using a machine learning model, but I encountered an exception:


The exception occurs when I use the machine learning model in the `OnAsioOutAudioAvailable` method. However, when I don't use the machine learning model, the code works fine.

Here is my code:

```csharp
public async Task Handle(ApplyEffectCommand command)
{
    _model = Helper.GetModel(Path.Combine(_modelDirectory, "my_model.json"), Path.Combine(_modelDirectory, "my_model.h5"));

    var waveformat = new WaveFormat(44100, 16, 1);
    var asioOut = new AsioOut("Audient USB Audio ASIO Driver");
    asioOut.InputChannelOffset = 0;
    asioOut.InitRecordAndPlayback(new BufferedWaveProvider(waveformat), 1, 44100);
    asioOut.AudioAvailable += OnAsioOutAudioAvailable;
    asioOut.Play();
}

private async void OnAsioOutAudioAvailable(object sender, AsioAudioAvailableEventArgs e)
{
    e.WrittenToOutputBuffers = true;
    var inputBuffers = e.InputBuffers;
    var outputBuffers = e.OutputBuffers;

    for (int channel = 0; channel < inputBuffers.Length; channel++)
    {
        // Get the input buffer pointer for the current channel
        IntPtr inputBufferPtr = inputBuffers[channel];

        // Get the output buffer pointer for the current channel
        IntPtr outputBufferPtr = outputBuffers[channel];

        // Calculate the number of samples in the buffer
        int bufferSizeSamples = e.SamplesPerBuffer;

        // Create a buffer to store the input data
        float[] inputBuffer = new float[bufferSizeSamples];

        // Marshal the input data from the IntPtr buffer
        Marshal.Copy(inputBufferPtr, inputBuffer, 0, bufferSizeSamples);

        // Apply effect from model
        var predicted = await Task.Run(() => _model.Predict(inputBuffer));

        Marshal.Copy(predicted.GetData<float>(), 0, outputBufferPtr, bufferSizeSamples);
    }
}

**Comment by markheath** [comment]
You can't await tasks in your ASIO callback. You have to run everything synchronously and fill the buffers as quickly as possible.

**Comment by NickHumphries** [comment]
Hi Mark,

Thanks very much for NAudio, I'm really enjoying using it in my simple music player.
I do have a problem though with ASIO. It keeps missing bits, sounding choppy. I guess that perhaps a message to fill the buffers is being missing, or filling the buffers is taking too long? WASAPII and WAVE work fine, even oversampled to 176.4 Khz, but even basic ASIO fails.
I'm using the latest Naudio and VS 2026 with Net 10.
Could you perhaps suggest where I can look to debug this?

Many thanks,

Nick.
